src/__tests__/clowder.test.ts

```diff
@@ -0,0 +1,585 @@
+      catToSelect: 'unvalidated',
+    });
+
+    expect(clowder.nItems.cat1).toBe(0);
+  });
+
+  it('should return undefined for next item if catToSelect = "unvalidated" and no unvalidated items remain', () => {
+    const clowderInput: ClowderInput = {
+      cats: {
+        cat1: { method: 'MLE', theta: 0.5 },
+      },
+      corpus: [
+        createMultiZetaStimulus('0', [createZetaCatMap([])]),
+        createMultiZetaStimulus('1', [createZetaCatMap(['cat1'])]),
+        createMultiZetaStimulus('2', [createZetaCatMap([])]),
+        createMultiZetaStimulus('3', [createZetaCatMap(['cat1'])]),
+      ],
+    };
+
+    const clowder = new Clowder(clowderInput);
+
+    const nextItem = clowder.updateCatAndGetNextItem({
+      catsToUpdate: ['cat1'],
+      items: [clowder.corpus[0], clowder.corpus[2]],
+      answers: [1, 1],
+      catToSelect: 'unvalidated',
+    });
+
+    expect(nextItem).toBeUndefined();
+  });
+
+  it('should correctly update ability estimates during the updateCatAndGetNextItem method', () => {
+    const originalTheta = clowder.cats.cat1.theta;
+    clowder.updateCatAndGetNextItem({
+      catToSelect: 'cat1',
+      catsToUpdate: ['cat1'],
+      items: [clowder.corpus[0]],
+      answers: [1],
+    });
+    expect(clowder.cats.cat1.theta).not.toBe(originalTheta);
+  });
+
+  it('should randomly choose between validated and unvalidated items if randomlySelectUnvalidated is true', () => {
+    const clowderInput: ClowderInput = {
+      cats: {
+        cat1: { method: 'MLE', theta: 0.5 },
+      },
+      corpus: [
+        createMultiZetaStimulus('0', [createZetaCatMap(['cat1'])]), // Validated item
+        createMultiZetaStimulus('1', [createZetaCatMap([])]), // Unvalidated item
+        createMultiZetaStimulus('2', [createZetaCatMap([])]), // Unvalidated item
+        createMultiZetaStimulus('3', [createZetaCatMap([])]), // Validated item
+      ],
+      randomSeed: 'randomSeed',
+    };
+    const clowder = new Clowder(clowderInput);
+
+    const nextItems = Array(20)
+      .fill('-1')
+      .map(() => {
+        return clowder.updateCatAndGetNextItem({
+          catToSelect: 'cat1',
+          randomlySelectUnvalidated: true,
+        });
+      });
+
+    const itemsId = nextItems.map((item) => item?.id);
+
+    expect(nextItems).toBeDefined();
+    expect(_uniq(itemsId)).toEqual(expect.arrayContaining(['0', '1', '2', '3'])); // Could be validated or unvalidated
+  });
+
+  it('should return undefined if no more items remain', () => {
+    const nextItem = clowder.updateCatAndGetNextItem({
+      catToSelect: 'cat1',
+      items: clowder.remainingItems,
+      answers: [1, 0, 1, 1, 0], // Exhaust all items
+    });
+
+    expect(nextItem).toBeUndefined();
+  });
+
+  it('can receive one item and answer as an input', () => {
+    const nextItem = clowder.updateCatAndGetNextItem({
+      catToSelect: 'cat1',
+      items: clowder.corpus[0],
+      answers: 1,
+    });
+    expect(nextItem).toBeDefined();
```

Comment:
what does "toBeDefined()" mean?

Reply:
This is a Jest "matcher" used for unit testing https://jestjs.io/docs/expect#tobedefined

Reply:
okay, thanks for explaining.